### PR TITLE
[mdspan] Add new port for kokkos/mdspan

### LINKED ISSUE #21183: [New Port Request] mdspan
Library name: mdspan

Library description: A non-owning multi-dimensional array reference type.

Source repository URL: https://github.com/kokkos/mdspan

Project homepage (if different from the source repository):

Anything else that is useful to know when adding (such as optional features the library may have that should be included):

### PR DESCRIPTION
**Pull request by jdthomas**
Adds new port for [mdspan](https://github.com/kokkos/mdspan)

- #### What does your PR fix?  
  Fixes #21183 , adds new port for mdspan

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all, No

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

### CODE REVIEW

File: ports/mdspan/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  OUT_SOURCE_PATH
+  SOURCE_PATH
+  REPO
+  kokkos/mdspan
+  REF
+  aced2cebd362a1e15830da030bd16748131d28bd # stable as of 2021-11-03
+  SHA512
+  a1950430be537497fb84c4a8c5e681cacead93512775098f38ea6c1a20b95d0f7110d9d0802fbdcf8ce3c40ade766cc697773f6ea6fcf8c363b3ebee55620f7c
+  HEAD_REF
+  stable)
```
Comment (NancyLi1013) [inline ports/mdspan/portfile.cmake:11]:
```suggestion
vcpkg_from_github(
    OUT_SOURCE_PATH SOURCE_PATH
    REPO kokkos/mdspan
    REF aced2cebd362a1e15830da030bd16748131d28bd # stable as of 2021-11-03
    SHA512 a1950430be537497fb84c4a8c5e681cacead93512775098f38ea6c1a20b95d0f7110d9d0802fbdcf8ce3c40ade766cc697773f6ea6fcf8c363b3ebee55620f7c
    HEAD_REF stable
)
```

File: ports/mdspan/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+vcpkg_cmake_configure(SOURCE_PATH ${SOURCE_PATH})
```
Comment (NancyLi1013) [inline ports/mdspan/portfile.cmake:13]:
```suggestion
vcpkg_cmake_configure(SOURCE_PATH "${SOURCE_PATH}")
```

File: ports/mdspan/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+vcpkg_cmake_install()
+
+vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/mdspan)
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib"
+     "${CURRENT_PACKAGES_DIR}/debug")
```
Comment (NancyLi1013) [inline ports/mdspan/portfile.cmake:20]:
```suggestion
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib" "${CURRENT_PACKAGES_DIR}/debug")
```

File: ports/mdspan/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+vcpkg_copy_pdbs()
+
+file(
+  INSTALL ${SOURCE_PATH}/LICENSE
+  DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT}
+  RENAME copyright)
```
Comment (NancyLi1013) [inline ports/mdspan/portfile.cmake:27]:
```suggestion
file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
```

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/mdspan/vcpkg.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "mdspan",
+  "version-date": "2021-11-03",
+  "description": "Provide a production-quality implementation of the P0009 proposal as written in preparation for the addition of mdspan to the c++ standard library.",
```
Comment (jdthomas) [inline ports/mdspan/vcpkg.json:4]:
```suggestion
  "description": "A non-owning multi-dimensional array reference type.",
```

Comment (jdthomas) [reply ports/mdspan/vcpkg.json:4]: @cjdb 's description is nicer than my copy-pasta from the README :)

Comment (cjdb) [reply ports/mdspan/vcpkg.json]: FWIW that was also a copypasta with an "A " prefixed :)

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
LGTM, thanks for adding this port. @jdthomas

**Comment by jdthomas** [comment]
> @jdthomas I haven't ever used vcpkg would it be possible to get some guidance on how we can make sure we don't break this integration?

@crtrott 
heh, I am new to this as well, so maybe there's someone better to answer any specific question :)
In the vcpkg repo there's a hard coded sha (aced2cebd362a1e15830da030bd16748131d28bd) that it pulls from the mdspan repo. I think the only maintenance you might be interested in is potentially updating that if mdspan gets updates. Since mdspan takes no external dependencies, I cannot imagine there being any breakage other than potentially while updating (and even then super unlikely unless someone adds another "port" that depends on mdspan).  ... but again, i am new to this too 🤷🏽‍♀️

**Pull request merged**